### PR TITLE
WIP: Fortran cmake

### PR DESCRIPTION
**Pull request by pclausen**
This is a work in progress pull request. Target is to use CMake for Fortran exercises. 

I have tried to create a mini-test framework, `config/TesterMain.f90`, which is added to all exercise directories. The `<exercise>_test.f90` contain the actual testing generated using a python script, `config/create_fortran_test.py` and then modified by hand.

I have got `hello-world, bob and hamming` up and running and travis build and test successfully. But I am new to github and exercism, so I would like to get some feedback before continuing down a path not wanted by the community. Please let me know if something is wrong in the place this pull request is done or with my fork, or if I can improve something. Thank you.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
-  - sudo apt-get install gfortran
-  - sudo apt-get install ruby-full
-
+dist: trusty
+sudo: false 
+#language: fortran
+language: C
+branches:
+  only:
+  - fortran-cmake
```
Comment (SaschaMann) [inline .travis.yml]: The tests should be run on all branches/PRs, so I think you can remove the `branches` section. If a branch modifies `.travis.yml`, Travis will run it only on that branch.

File: .gitignore [hunk 2]
```diff
@@ -2,3 +2,18 @@
 .DS_Store
```
Comment (kotp) [inline .gitignore]: This line should go away.  This is user environment and doesn't belong in the project.

File: docs/INSTALLATION.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+* a modern Fortran compiler
+* the CMake cross-platform build system
+
+### Prerequisite: A Modern Fortran Compiler 
+
+This language track requires a compiler with [Fortran 2003](https://en.wikipedia.org/wiki/Fortran#Fortran_2003)
+support. All major compilers released in the last few years should
+be compatible, so as long as you are on a fairly recent version you should be fine.
+
+The following will describe installation of [GNU Fortran](https://gcc.gnu.org/fortran/) or GFortran. Other fortran compilers are listed [here](https://en.wikipedia.org/wiki/List_of_compilers#Fortran_compilers). Intel Fortran is a popular proprietary choice for high performance applications. Most exercises will work with Intel Fortran, but are only tested with GNU Fortran so your milage may vary.
```
Comment (kotp) [inline docs/INSTALLATION.md]:
```suggestion
The following describes the installation of [GNU Fortran](https://gcc.gnu.org/fortran/) or GFortran.
Other fortran compilers are listed [here]https://en.wikipedia.org/wiki/List_of_compilers#Fortran_compilers).
Intel Fortran is a popular proprietary choice for high performance applications.
Most exercises will work with Intel Fortran, but are only tested with GNU Fortran so your milage may vary. 
```

If we do one sentence per line, the corrections (diffs) won't be as noisy, we hope.  Up to you, though how we do this, as "Chief Maintainer".  I am not opposed to 80 character (or thereabout) line limit, but line per sentence has its benefits over 80 characters per line.

File: docs/INSTALLATION.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+* fails the build if the any tests fail
+
+Using CMake allows exercism to provide a cross-platform build script that
+can generate project files for integrated development environments like
+Visual Studio and Xcode.  This allows you to focus on the problem and
+not worry about setting up a build for each exercise.
+
+Getting a portable build isn't easy and requires access to many kinds of
+systems.  If you encounter any problems with the supplied CMake recipe,
+please [report the issue](https://github.com/exercism/cpp/issues) so we can
```
[the remark this answers is not in the document]
Comment (kotp) [reply docs/INSTALLATION.md]:
```suggestion
please [report the issue](https://github.com/exercism/fortran/issues) so we can
```

File: docs/INSTALLATION.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
```
[the remark this answers is not in the document]
Comment (kotp) [reply docs/INSTALLATION.md]:
```suggestion
please [report the issue](https://github.com/exercism/fortran/issues) so we can
```

**Comment by pclausen** [comment]
Thanks for the feedback. Some answers:

@SaschaMann 
> Is there a commonly used unit testing library for Fortran? Previously, FUnit was used, are there reasons against using it? (To prevent misunderstandings: I'm just curious, this isn't a disapproval at all)

I had a hard look at FUnit and did not find that it brought many advantages. On the disadvantage side, most Fortran users I know of would not have Ruby and even though I have some linux and admin experience I thought it was too much hassle to get Ruby installed first. In my company we have also looked for a good Unittesting framework for Fortran but ended up writing our own stuff similar to TesterMain.f90 which I wrote based on what I felt I needed. 

If you are missing certain features in such a Test-module I would be very interested as I am not very used to Test Frameworks other than Python [unittest](https://docs.python.org/3/library/unittest.html) which I like because its fairly simple to use and minimal overhead.

> Is it somehow possible to install a library that contains TesterMain instead of including it in every exercise directory?

Yes, that just needs a bit more CMake magic to happen. I would do this in a later change.

**Comment by pclausen** [comment]
@kotp 
>  Think of these three categories of changes:
> - Documentation
> - Configuration
> -  Exercises


I am fine with this kind of separation, but again I dont know how to do this in git. Would it be OK to incorporate the changes you and others have suggested into this single PR ? Or, if you could explain to how I should split the changes into different PRs/branches/.... I would try to go that route, but I am very uncertain to what the best way of handling this is.

**Comment by SaschaMann** [comment]
> If you are missing certain features in such a Test-module I would be very interested as I am not very used to Test Frameworks other than Python unittest which I like because its fairly simple to use and minimal overhead.

A feature that tests that a certain error is thrown for given arguments is missing in this, as far as I can see. I don't know how errors are handled in Fortran, but that might be something to add in the future if an exercise requires it.

**Comment by pclausen** [comment]
This work was put into pull requests #36-#62.

**Pull request closed without merging**